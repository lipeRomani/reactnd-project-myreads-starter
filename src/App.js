import React from 'react'
import {Route} from 'react-router-dom';
import * as BooksAPI from './BooksAPI'
import './App.css'
import BookListBox from './BookListBox';
import SearchBox from './SearchBox';

class BooksApp extends React.Component {
  state = {
    myBooks: [],
    booksFound : [],
    searchQuery : "" 
  }

  componentDidMount() {
    BooksAPI
      .getAll()
      .then(myBooks => this.setState({myBooks}));
  }

  changeShelf = (book, shelfName) => {
    return BooksAPI
      .update(book, shelfName)
      .then(result => BooksAPI.getAll())
      .then(myBooks => {
        this.setState({myBooks});
        if (this.state.searchQuery.length > 0) {
          this.searchBook(this.state.searchQuery);
        }
      });
  }

  searchBook = (query) => {
    this.setState({searchQuery : query});

    if (query.length === 0) {
      this.setState({booksFound : []});
    }

      BooksAPI
          .search(query)
          .then(booksFound => {
            if (!Array.isArray(booksFound)) {
              this.setState({booksFound : []})
              return [];
            }

            let newBooksFound =  booksFound.map((bookFound) => {
              for (let book of this.state.myBooks) {
                if (book.id === bookFound.id) {
                  return book;
                }
              } 
              return bookFound;
            });

            this.setState({booksFound : newBooksFound})
          })
    
  }

  render() {
    return (
      <div className="app">
        <Route
          exact
          path="/"
          render={() => {
          return <BookListBox books={this.state.myBooks} changeShelf={this.changeShelf} />;
        }}/>

        <Route
          path="/search"
          render={() => {
          return <SearchBox 
              searchBook={this.searchBook} 
              booksFound={this.state.booksFound} 
              query={this.state.searchQuery} 
              changeShelf={this.changeShelf}
              />;
        }}/>
      </div>
    )
  }
}

export default BooksApp
