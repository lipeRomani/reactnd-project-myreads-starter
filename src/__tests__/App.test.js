import React from 'react';
import App from '../App';
import {shallow} from 'enzyme';

/** 
 This course is not designed to teach Test Driven Development. 
 Feel free to use this file to test your application, but it 
 is not required.
**/

it('renders without crashing', () => {
  const wrapper = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
})



