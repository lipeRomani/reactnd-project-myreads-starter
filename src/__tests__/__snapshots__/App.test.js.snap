exports[`test renders without crashing 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="app">
    <Route
        exact={true}
        path="/"
        render={[Function]} />
    <Route
        path="/search"
        render={[Function]} />
</div>,
  "nodes": Array [
    <div
      className="app">
      <Route
            exact={true}
            path="/"
            render={[Function]} />
      <Route
            path="/search"
            render={[Function]} />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <BooksApp />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": BooksApp {
        "_reactInternalInstance": [Circular],
        "changeShelf": [Function],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "searchBook": [Function],
        "state": Object {
          "booksFound": Array [],
          "myBooks": Array [],
          "searchQuery": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="app">
          <Route
                    exact={true}
                    path="/"
                    render={[Function]} />
          <Route
                    path="/search"
                    render={[Function]} />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="app">
          <Route
                    exact={true}
                    path="/"
                    render={[Function]} />
          <Route
                    path="/search"
                    render={[Function]} />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <BooksApp />,
}
`;
